File: Cpp/1475_Final_Prices_With_a_Special_Discount_in_a_Shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        vector<int> result(prices.size(), 0);
        stack<int> s;
        for(int i=prices.size()-1; i>=0; --i){
            while (!s.empty() && s.top() > prices[i]) {
                s.pop();
            }
            result[i] = s.empty() ? prices[i] : prices[i] - s.top();
            s.push(prices[i]);
        }
        
        return result;
    }
};

File: Kotlin/1475_Final_Prices_With_a_Special_Discount_in_a_Shop.kt
class Solution {

    fun finalPrices(prices: IntArray): IntArray {
        val result = IntArray(prices.size)
        val s = Stack<Int>()
        
        for(i in prices.lastIndex downTo 0) {
            while(!s.isEmpty() && prices[s.peek()] > prices[i]) {
                s.pop()
            }
            result[i] = if(s.isEmpty()) prices[i] else prices[i] - prices[s.peek()]
            s.push(i)
        }
        
        return result
    }
}